Scenarios documentation (#49)
loadimpact/k6-docs: src/data/markdown/docs/01 guides/02 Using k6/14 Executors.md

---
title: Executors
excerpt: ''
hideFromSidebar: false
---

k6 v0.27.0 introduces the concept of _executors_: configurable modes of executing a
JavaScript function that can model diverse traffic scenarios in load tests.

Multiple executors can be used in the same script, they can be scheduled to run
parallel or in sequence, and each executor can run a different JS function with
different environment variables and tags, so this brings a lot of flexibility with
organizing and modeling testing scenarios.

Existing global execution options such as `vus`, `duration` and `stages` were
formalized into standalone executors, while new executors were added to support more
advanced modes of execution. A major benefit of this restructuring is that support
for new testing scenarios can be added relatively easily to k6, making it very

> extensible.

Reviewer: Do you mean new "executors"?

Author: You mean instead of "new testing scenarios"? Could be, but I wanted to focus on the end-user benefit (being able to model new testing scenarios _with_ new executors) rather than just the addition of a new executor, which is an internal benefit. It's subtle, but I prefer the way it's currently worded. :)

Author: I removed this paragraph in 17b520a. It sounded too changelog-y and it's an internal/development benefit, no need to mention it here.

> ### ⚠  Backwards compatibility
>
> Note that pre-v0.27.0 scripts and options should continue to work
> the same as before (with a few minor breaking changes, as mentioned
> in the [release notes](https://github.com/loadimpact/k6/releases/tag/v0.27.0))
> but please [create a bug issue](https://github.com/loadimpact/k6/issues/new?labels=bug&template=bug_report.md)
> if that's not the case.

## Executor types

### Shared iterations

> A fixed number of iterations are "shared" by all VUs, and the test ends once all

Reviewer: ```suggestion
A fixed number of iterations are "shared" between a number of VUs, and the test ends once all
```
Also we should mention that "faster" VUs will execute more iterations

Author: Resolved in f99fbf3.

iterations are executed. This executor is equivalent to the global `vus` and

> `iterations` options.

Author: We should mention the unique identifier name for each executor here (`shared-iterations` in this case), even though it's easy to deduce. Any ideas where? Maybe in parenthesis after the header?

Author: Actually, it might be better if there's a small "Examples" section under each executor, so that it's clear what the unique identifier is. Initially I opted against this as these examples will essentially only differ in the `options`, which are already documented, so it seems redundant.

But after more consideration, I'd rather have these small examples there, and rename the current standalone "Examples" to "Advanced examples" where 2 or more executors are used at the same time and to showcase other more complex behavior.

Thoughts?

| `startVUs`         | integer | Number of VUs to run at test start.                                           | `1`     |
| `stages`           | array   | Array of objects that specify the target number of VUs to ramp up or down to. | `[]`    |
| `gracefulRampDown` | string  | Time to wait for iterations to finish before starting new VUs.                | `"30s"` |

### Externally controlled

Control and scale execution at runtime via k6's REST API or the CLI.

This executor formalizes what was [previously possible globally](https://k6.io/blog/how-to-control-a-live-k6-test),
and using it is required in order to use the `pause`, `resume`, and `scale` CLI commands.

| Option     | Type    | Description                                         | Default |
|------------|---------|-----------------------------------------------------|---------|
| `vus`      | integer | Number of VUs to run concurrently.                  | -       |
| `duration` | string  | Total test duration.                                | -       |
| `maxVUs`   | integer | Maximum number of VUs to allow during the test run. | -       |

> 

Reviewer: This should be last ;)

Author: :+1: Resolved in f99fbf3.

### Per VU iterations

Each VU executes a fixed number of iterations.

| Option        | Type    | Description                                                                    | Default |
|---------------|---------|--------------------------------------------------------------------------------|---------|
| `vus`         | integer | Number of VUs to run concurrently.                                             | `1`     |
| `iterations`  | integer | Number of script iterations to execute with each VU.                           | `1`     |

> | `maxDuration` | string  | Maximum test duration before it's forcibly stopped (excluding `gracefulStop`). | `"10m"` |

Reviewer: This should be just after the `shared iterations` IMO

Author: :+1: Resolved in f99fbf3.

### Constant arrival rate

A fixed number of iterations are executed in a specified period of time. This allows
k6 to dynamically change the amount of VUs during a test run to achieve the specified
amount of iterations per period, which is useful for a more accurate representation

> of RPS, for example. See [issue #550](https://github.com/loadimpact/k6/issues/550) for details.

Author: I reworded your suggestion a bit :) Let me know if f99fbf3 covers it.

Also, I guess the answer to my question [here](https://github.com/loadimpact/k6-docs/pull/49/files#diff-04ed84b1653e21012160c13d823f4a66R105) is that we should link to the arrival rate section you suggested in this same doc, instead of to #550.

| `maxVUs`          | integer | Maximum number of VUs to allow during the test run.                                     | -       |

## Shared configuration

All executors share a base configuration, so the following options can be used for all:

| Option         | Type   | Description                                                                      | Default     |
|----------------|--------|----------------------------------------------------------------------------------|-------------|
| `startTime`    | string | Time offset since test start this scenario should begin execution.               | `"0s"`      |
| `gracefulStop` | string | Time to wait for iterations to finish executing before stopping them forcefully. | `"30s"`     |
| `exec`         | string | Name of exported JS function to execute.                                         | `"default"` |
| `env`          | object | Environment variables specific to this scenario.                                 | `{}`        |
| `tags`         | object | [Tags](/using-k6/tags-and-groups) specific to this scenario.                     | `{}`        |

> 

Author: I considered that, but then felt it was better to introduce all executors first, otherwise these options could be confusing.

But you might be right, I'll move it to the top. :)

> 

Author: Resolved in f99fbf3.

## Examples

- Execute a constant 200 RPS for 1 minute, allowing k6 to dynamically schedule up to 100 VUs.

Note that in order to reliably achieve a fixed request rate, it's recommended to keep
the function being executed very simple, with preferably only a single request call,

> and no additional processing or `sleep()` calls.

Author: :+1: for the separate arrival rate section.

Would you or @robingustafsson volunteer for that? :smile: I could copy or paraphrase what Robin wrote in https://github.com/loadimpact/k6/issues/550, but you guys have a better grasp of this and will probably explain it better.

> and no additional processing or `sleep()` calls.

Author: I added a placeholder for the arrival rate section in 95f1187.

<div class="code-group" data-props='{"labels": [ "constant-arr-rate.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    constant_arr_rate: {  // arbitrary scenario name
      executor: 'constant-arrival-rate',
      rate: 200,
      timeUnit: '1s',
      duration: '1m',
      preAllocatedVUs: 50,

>       maxVUs: 100,

Author: Hah I thought this one is obviously arbitrary but maybe not :)

Let me know if f99fbf3 is an improvement.

- Execute a variable RPS test, starting at 50, ramping up to 200 and then back to 0,
  over a 1 minute period:

<div class="code-group" data-props='{"labels": [ "ramping-arr-rate.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    ramping_arrival: {  // arbitrary scenario name
      executor: 'ramping-arrival-rate',
      startRate: 50,
      timeUnit: '1s',
      preAllocatedVUs: 50,
      maxVUs: 100,
      stages: [
        { target: 200, duration: '30s' },
        { target: 0, duration: '30s' },
      ],
    }
  }
};

export default function() {
  http.get('https://test.k6.io/');
}
```

</div>

- Run as many iterations as possible with 50 VUs for 30s, and then run 100 iterations
  per VU of another scenario.

Note the use of `startTime`, and different `exec` functions for each scenario.

<div class="code-group" data-props='{"labels": [ "multiple-scenarios.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    const_vus: {
      executor: 'constant-vus',
      exec: 'scenario1',
      vus: 50,

>       duration: '30s',

Author: I would rather not introduce `gracefulStop` here since this example already showcases a couple of features, and `gracefulStop` would need a more thorough explanation. And it wasn't meant to stop at _exactly_ 30s. :)

So let's use `gracefulStop` in a separate example in that new gracefulStop/RampDown section you mentioned.

>       duration: '30s',

Author: See the `gracefulStop` section added in 72d17e4.

      startTime: '30s',
      maxDuration: '1m',
    },
  }
};

export function scenario1() {
  http.get('https://test.k6.io/', { tags: { scenario: '1' }} );
}

export function scenario2() {
  http.get('https://test.k6.io/', { tags: { scenario: '2' }} );
}
```

</div>

- Use different environment variables and tags per scenario.

In the previous example we set tags on individual HTTP request metrics, but this
can also be done per scenario, which would apply them to other
[taggable](https://k6.io/docs/using-k6/tags-and-groups#tags) objects as well.

<div class="code-group" data-props='{"labels": [ "multiple-scenarios-env-tags.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';
import { fail } from 'k6';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    const_vus: {
      executor: 'constant-vus',
      exec: 'scenario1',
      vus: 50,
      duration: '30s',
      tags: { scenario: '1' },
      env: { SCENARIO: '1' },
    },
    per_vu_iters: {
      executor: 'per-vu-iterations',
      exec: 'scenario2',
      vus: 50,
      iterations: 100,
      startTime: '30s',
      maxDuration: '1m',
      tags: { scenario: '2' },

>       env: { SCENARIO: '2' },

Reviewer: if we are going to be setting those in k6 it's probably not a good idea to give examples like that ;)

Author: Hah true, though I wrote this before the default `scenario` tag feature, and well, because of lack of imagination :)

I changed it slightly in f99fbf3, but let me know if you have a better idea.

---

> title: Executors

Reviewer 2: I think we should either call this guide "Scenarios" and rework the page a little bit to first talk about scenarios as a concept and explaining multi-scenario tests before going into the details of each executor which is a setting for a scenario, or we should introduce a new guide called "Scenarios" that is introduced before this guide that talks about scenarios as a concept etc. linking to this guide for more details.

Author: Makes sense. The feature as a whole is "scenarios", and "executors" are a (crucial) part that make them possible, so we should mention and list them, but not make it the focal point of the article.

I'll see what I can add in the intro, and we can expand it to a standalone article if needed.

Author: @robingustafsson Let me know if 17b520a is sufficient for what you had in mind. I think this is a gentler introduction, focusing on the benefits of scenarios and a general configuration example before going into details with executors. I'm not sure what else we can mention there, so maybe it doesn't need a separate article?

that arguments to `scale` to change the amount of active or maximum VUs only affect
the externally controlled executor.

| Option     | Type    | Description                                         | Default |
|------------|---------|-----------------------------------------------------|---------|
| `vus`      | integer | Number of VUs to run concurrently.                  | -       |
| `maxVUs`   | integer | Maximum number of VUs to allow during the test run. | -       |
| `duration` | string  | Total test duration.                                | -       |

#### Examples

- Execute a test run controllable at runtime, starting with 0 VUs up to a maximum of
  50, and a total duration of 10 minutes:

<div class="code-group" data-props='{"labels": [ "externally-controlled.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    contacts: {
      executor: 'externally-controlled',
      vus: 0,
      maxVUs: 50,
      duration: '10m',
    }
  }
};

export default function() {
  http.get('https://test.k6.io/contacts.php');
}
```

</div>

## Arrival rate

> TODO (expand on https://github.com/loadimpact/k6/issues/550)

Reviewer 2: I can take a stab at writing this section intro if you want.

Author: Thanks, I think that would be best as you have more knowledge about the topic. I'd just be paraphrasing :)

Reviewer 2: Ok, have pushed a first draft of content for the arrival rate section now :)

Author: That looks great, the examples make it very easy to understand. Just fixed a couple of typos :)

## Graceful stop and ramp down

In versions before v0.27.0, k6 would interrupt any iterations being executed if the
test is done or when ramping down VUs when using the `stages` option. This behavior
could cause skewed metrics and wasn't user configurable.

In v0.27.0 a new option is introduced for all executors (except externally-controlled):
`gracefulStop`. With a default value of `30s`, it specifies the time k6 should wait
for iterations to complete before forcefully interrupting them.

A similar option exists for the ramping VUs executor: `gracefulRampDown`. This
specifies the time k6 should wait for any iterations in progress to finish before

> VUs are returned to the global pool during the ramp down period defined in `stages`.

Reviewer 2: I'd change "*the* ramp down period" to "*a* ramp down period" as ramp downs aren't mandatory and there can be more than one :)

Author: Resolved in 26911c0.

---
title: Executors
excerpt: ''
hideFromSidebar: false
---

k6 v0.27.0 introduces the concept of _executors_: configurable modes of executing a
JavaScript function that can model diverse traffic scenarios in load tests.

> Multiple executors can be used in the same script, they can be scheduled to run

Author: I reworked the intro somewhat to focus on scenarios, and now it's "run in sequence or parallel". Not sure if a second "in" is needed before "parallel", but DRY? :smile:

| `maxDuration` | string  | Maximum test duration before it's forcibly stopped (excluding `gracefulStop`). | `"10m"` |

#### Examples

- Execute 200 total iterations shared by 10 VUs with a maximum duration of 10 seconds:

<div class="code-group" data-props='{"labels": [ "shared-iters.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    contacts: {  // arbitrary scenario name
      executor: 'shared-iterations',
      vus: 10,
      iterations: 200,
      maxDuration: '10s',
    }
  }
};

export default function() {
  http.get('https://test.k6.io/contacts.php');
}
```

</div>

### Per VU iterations

Each VU executes an exact number of iterations.

| Option        | Type    | Description                                                                    | Default |
|---------------|---------|--------------------------------------------------------------------------------|---------|
| `vus`         | integer | Number of VUs to run concurrently.                                             | `1`     |

> | `iterations`  | integer | Number of script iterations to execute with each VU.                           | `1`     |

Reviewer: ```suggestion
| `iterations`  | integer | Number of `exec` function iterations to be executed by each VU.                           | `1`     |
```

Author: Resolved in 2d07d72.

#### Examples

- Run a two-stage test, ramping up from 0 to 100 VUs for 5 seconds, and down to 0 VUs
  for 5 seconds:

<div class="code-group" data-props='{"labels": [ "ramping-vus.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    contacts: {
      executor: 'ramping-vus',
      startVUs: 0,
      stages: [
        { duration: '5s', target: 100 },
        { duration: '5s', target: 0 },
      ],
      gracefulRampDown: '0s',
    }
  }
};

export default function() {
  http.get('https://test.k6.io/contacts.php');
}
```

</div>

Note the setting of `gracefulRampDown` to 0 seconds, which could cause some
iterations to be interrupted during the ramp down stage.

### Constant arrival rate

A fixed number of iterations are executed in a specified period of time.
Since iteration execution time can vary because of test logic or the
system-under-test responding more slowly, this executor will try to compensate
by running a variable number of VUs--including initializing more in the middle
of the test--in order to meet the configured iteration rate. This approach is

> useful for a more accurate representation of RPS, for example.

Author: It was supposed to be an em dash, to represent an interjection of sorts, and some Markdown engines render it properly, but not Gatsby apparently. So I'll change it to the HTML codes, unless you think we should reword the sentence to avoid it.

> useful for a more accurate representation of RPS, for example.

Author: Resolved in 2d07d72?

> useful for a more accurate representation of RPS, for example.

Reviewer: > It was supposed to be an em dash,
I didn't know this :D, but yeah if it doesn't work in Gatsby ...

  // and we can thus expect to get 60 iterations completed
  // for the full 1m test duration.
  http.get('http://httpbin.test.k6.io/delay/10');
}
```

</div>

Running this script would result in something like:

```
running (1m09.3s), 000/011 VUs, 60 complete and 0 interrupted iterations
open_model ✓ [======================================] 011/011 VUs  1m0s  1 iters/s
```

Compared to the first example of the closed model, in this open model example we can see
that the VU iteration arrival rate is now decoupled from the iteration duration.
The response times of the target system are no longer influencing the load being
put on the target system.

## Graceful stop and ramp down

In versions before v0.27.0, k6 would interrupt any iterations being executed if the
test is done or when ramping down VUs when using the `stages` option. This behavior
could cause skewed metrics and wasn't user configurable.

In v0.27.0 a new option is introduced for all executors (except externally-controlled):
`gracefulStop`. With a default value of `30s`, it specifies the time k6 should wait
for iterations to complete before forcefully interrupting them.

> A similar option exists for the ramping VUs executor: `gracefulRampDown`. This

Author: :+1: Resolved in 2d07d72.

In the previous example we set tags on individual HTTP request metrics, but this
can also be done per scenario, which would apply them to other
[taggable](https://k6.io/docs/using-k6/tags-and-groups#tags) objects as well.

<div class="code-group" data-props='{"labels": [ "multiple-scenarios-env-tags.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';
import { fail } from 'k6';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    contacts: {
      executor: 'constant-vus',
      exec: 'contacts',
      vus: 50,
      duration: '30s',
      tags: { my_custom_tag: 'contacts' },
      env: { MYVAR: 'contacts' },
    },
    news: {
      executor: 'per-vu-iterations',
      exec: 'news',
      vus: 50,
      iterations: 100,
      startTime: '30s',
      maxDuration: '1m',
      tags: { my_custom_tag: 'news' },
      env: { MYVAR: 'news' },
    },
  }
};

export function contacts() {
  if (__ENV.MYVAR != 'contacts') fail();
  http.get('https://test.k6.io/contacts.php');
}

export function news() {
  if (__ENV.MYVAR != 'news') fail();
  http.get('https://test.k6.io/news.php');
}
```

</div>

> Note that by default a `scenario` tag with the name of the scenario as value is

Author: I added it to the default list of `--system-tags` on the Options page. Or do you mean somewhere else?

> Note that by default a `scenario` tag with the name of the scenario as value is

Reviewer: in the [tags and groups](https://github.com/loadimpact/k6-docs/blob/55e7c2729755983496c17bab8cdaedf9e689bd90/src/data/markdown/docs/01%20guides/02%20Using%20k6/08%20Tags%20and%20Groups.md) ;)

> Note that by default a `scenario` tag with the name of the scenario as value is

Author: Done, see 8c0ee99.
